lib/common.js

```diff
@@ -66,44 +66,40 @@ function addRoute (interceptor, url, port, onMessage) {
   return true
 }
 
-function removeRoute (interceptor, thread) {
+async function removeRoute (interceptor, thread) {
   const routes = interceptor[kRoutes]
-
-  const allPorts = new Set(
-    Array.from(routes.values())
-      .map(r => r.ports)
-      .flat()
-  )
-
-  const allUrls = Array.from(routes.keys())
+  const removedPorts = []
 
   // Remove all routes for the thread
-  for (const url of allUrls) {
-    const roundRobin = routes.get(url)
+  for (const [url, roundRobin] of routes) {
     const port = roundRobin.findByThreadId(thread)
-
-    if (!port) {
-      continue
-    }
+    if (!port) continue
 
     roundRobin.delete(port)
-    allPorts.delete(port)
+    removedPorts.push(port)
 
     if (roundRobin.length === 0) {
       routes.delete(url)
     }
   }
 
-  // Return all the ports that are still alive
-  return allPorts
+  for (const port of removedPorts) {
+    const inflightReqs = port[kInflightOutgoing]
+    while (inflightReqs.size > 0) {
+      await new Promise(resolve => setImmediate(resolve))
```

Comment:
why are we adding `m*n` macroticks here? Can't we use an event?

Reply:
I replaced it with a Map wrapper with `drained` method.